test/test-try-commit.js

```diff
@@ -0,0 +1,523 @@
+	t.plan(6);
+	return ava(a => {
+		t.is(a.title, 'test');
+
+		return Promise.all([
+			a.try(b => {
+				t.is(b.title, 'test (attempt 1)');
+
+				return Promise.all([
+					b.try(c => t.is(c.title, 'test (attempt 1) (attempt 1)')),
+					b.try(c => t.is(c.title, 'test (attempt 1) (attempt 2)'))
+				]);
+			}),
+			a.try(b => t.is(b.title, 'test (attempt 2)')),
+			a.try(b => t.is(b.title, 'test (attempt 3)'))
+		]);
+	}).run();
+});
+
+test('try-commit does not fail when calling commit twice', t => {
+	return ava(a => {
+		return a.try(b => b.pass()).then(res => {
+			res.commit();
+			res.commit();
+		});
+	}).run().then(result => {
+		t.true(result.passed);
+		t.false(result.error);
+	});
+});
+
+test('try-commit does not fail when calling discard twice', t => {
+	return ava(a => {
+		return a.try(b => b.pass()).then(res => {
+			res.discard();
+			res.discard();
+		});
+	}).run().then(result => {
+		t.false(result.passed);
+		t.ok(result.error);
+		t.match(result.error.message, /Test finished without running any assertions/);
+		t.is(result.error.name, 'Error');
+	});
+});
+
+test('try-commit allows planning inside the try', t => {
+	return ava(a => {
+		return a.try(b => {
+			b.plan(3);
+
+			b.pass();
+			b.pass();
+			b.pass();
+		}).then(res => {
+			t.true(res.passed);
+			res.commit();
+		});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit fails when plan is not reached inside the try', t => {
+	return ava(a => {
+		return a.try(b => {
+			b.plan(3);
+
+			b.pass();
+			b.pass();
+		}).then(res => {
+			t.false(res.passed);
+			res.commit();
+		});
+	}).run().then(result => {
+		t.false(result.passed);
+	});
+});
+
+test('try-commit passes with failing test', t => {
+	return ava.failing(a => {
+		return a
+			.try(b => b.fail())
+			.then(res => {
+				t.false(res.passed);
+				t.ok(res.errors);
+				t.is(res.errors.length, 1);
+				const error = res.errors[0];
+				t.match(error.message, /Test failed via `t\.fail\(\)`/);
+				t.is(error.name, 'AssertionError');
+				res.commit();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit works with callback test', t => {
+	return ava.cb(a => {
+		a
+			.try(b => b.pass())
+			.then(res => {
+				res.commit();
+				a.end();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit works with failing callback test', t => {
+	return ava.cb.failing(a => {
+		a
+			.try(b => b.fail())
+			.then(res => {
+				t.false(res.passed);
+				t.ok(res.errors);
+				t.is(res.errors.length, 1);
+				const error = res.errors[0];
+				t.match(error.message, /Test failed via `t\.fail\(\)`/);
+				t.is(error.name, 'AssertionError');
+				res.commit();
+			})
+			.then(() => {
+				a.end();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit does not allow to use .end() in attempt when parent is callback test', t => {
```

Comment:
We should also test what happens if you use `.end()` when the parent is a regular test.